dev/core#4088 Convert api4 helper functionality to a trait & make available

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Convert api4 helper functionality to a trait & make available

First - ask jenkins....

Before
----------------------------------------
_What is the old user-interface or technical-contract (as appropriate)?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

After
----------------------------------------
_What changed? What is new old user-interface or technical-contract?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

Technical Details
----------------------------------------
_If the PR involves technical details/changes/considerations which would not be manifest to a casual developer skimming the above sections, please describe the details here._

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

eileenmcnaughton committed:
Convert api4 helper functionality to a trait & make available

totten commented on Civi/Test/Api4TestTrait.php:
```diff
@@ -0,0 +1,271 @@
+    unset($record);
+    $saved = civicrm_api4($entityName, 'save', $saveParams);
+    foreach ($saved as $item) {
+      $this->testRecords[] = [$entityName, [[$idField, '=', $item[$idField]]]];
```
Can this `$testRecords` property be extracted too?

eileenmcnaughton replied:
as in `setTestRecords` & `addTestRecord`?

totten replied:
As in `public $testRecords;` or `protected $testRecords;`

eileenmcnaughton replied:
hmm yeah ok - I'll add the getters & setters too if I'm going that

eileenmcnaughton replied:
Actually - it can be private so I don't need them

eileenmcnaughton commented on Civi/Test/Api4TestTrait.php:
```diff
@@ -0,0 +1,306 @@
+   * @param int $len
+   * @return string
+   */
+  protected function randomLetters(int $len = 10): string {
```
this is called from one child class - tbh I'd rather make it private & have that class do it's own thing

eileenmcnaughton commented on Civi/Test/Api4TestTrait.php:
```diff
@@ -0,0 +1,306 @@
+   * @return array
+   * @throws \CRM_Core_Exception
+   */
+  public function getRequiredValuesToCreate(string $entity, array &$values = []): array {
```
this is protected as the Conformance calls it directly

eileenmcnaughton replied:
we return `$values` - maybe we can not pass-by-ref as well....

eileenmcnaughton committed:
Clarify function signature

It was doing both pass-by-ref and return to give back the value - this
settles on using return.

totten commented on Civi/Test/Api4TestTrait.php:
```diff
@@ -0,0 +1,306 @@
+ *
+ * @package Civi\Test
+ *
+ * This trait defines a number of helper functions for testing APIv4.
```
I'd find it easier to understand with some kind of pointer about how to use the class. (*Which of the 8 helpers should I care about?*) Maybe like:

```suggestion
 * Define helpers for generating, tracking, and cleaning test-data with APIv4. Specifically:
 * - `createTestRecord()` will autofill a new record with example data.
 * - `deleteTestRecords()` will batch-delete all the test-data created by `createTestRecord()`
```

eileenmcnaughton replied:
Yep - I think those might be the only ones we would encourage people to use so that's probably right - I'm trying to privatise as much else as I can

eileenmcnaughton replied:
@totten I've updated this but what I found is there are 2 functions

`createTestRecord()`
`saveTestRecords()`

I suppose that's consistent with apiv4 but it feels like they should use the same verb - given it would be work to change I am VERY happy to be convinced otherwise

colemanw replied:
Yea, they both use the APIv4 `save` action, so changing to `saveTestRecord()` would be more consistent all-round.

eileenmcnaughton committed:
Move over testRecords property

seamuslee001 commented:
@eileenmcnaughton once you have resolved the comments from Tim I would say this is MOP

eileenmcnaughton committed:
Use Trait directly from managed entity test

I'm not sure if the test fail was an intermittent or related but this seems cleaner

eileenmcnaughton commented:
@seamuslee001 thanks - 

I got one fail - not sure if it is co-incidental or related - trying again

api\v4\Entity\ManagedEntityTest::testRevertSavedSearch
Failed asserting that '2023-01-24 22:33:06' is equal to '2023-01-24 22:33:07' or is greater than '2023-01-24 22:33:07'.

/home/jenkins/bknix-dfl/build/build-1/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Entity/ManagedEntityTest.php:119
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:1721

eileenmcnaughton commented:
Hmm - well I made it worse there! Fortunately local fails too so I'll find & fix

eileenmcnaughton committed:
Cleanup comments, remove unnecessary public function

eileenmcnaughton commented:
@seamuslee001 @totten it is passing now - the only hesitation I have is https://github.com/civicrm/civicrm-core/pull/25412#discussion_r1086149835 - but that is also ok for a follow up if we dither

colemanw commented:
@eileenmcnaughton +1 on changing `createTestRecord` to `saveTestRecord`, but otherwise my only beef with this PR is that:

- **Before:** saved test records would be automatically cleaned up during teardown; the author of the test doesn't have to think about it.
- **After:** now the trait provides a cleanup function, but no guidance on how to use it.

I've seen a lot of unit-test authors struggle with the concept of cleanup. A lot of people try to clean stuff (if they do it at all) by putting some `delete` calls at the end of their test, which makes sense to a procedural thought-process but doesn't work for unit tests. I worry about people calling `deleteTestRecords` from within their tests, or forgetting to call it at all.
My recommendation would be to add a `tearDown` function to this trait. That will cause annoying collisions with any `tearDown` function implemented in a class using this trait, which IMO is a good thing because it forces the dev to think about cleanup! To resolve the conflict they can simply do:

```
use  Api4TestTrait {
  tearDown as testRecordTearDown;
}

public function tearDown() {
  $this->testRecordTearDown();
  // other stuff
}
```

eileenmcnaughton commented:
@colemanw ok - I think both the rename & the tearDown fix are going to each involve touching quite a few classes so I think it would be helpful if we can merge this first

totten commented:
Yeah, I agree the tear-down protocol is the fiddly thing. Let's verbalize some options:

1. Continue emphasizing `Api4TestBase` as the main way to use this. Do whatever you need to make it seem more official. `Api4TestTrait` is the more advanced tool (for someone wiring-up their own base-class).
2. Encourage test-classes to resolve function conflicts (`use  Api4TestTrait {  tearDown as testRecordTearDown; }`).
3. Move the cleanup into `CiviTestListener`. If a class includes `Api4TestTrait`, then automatically call `deleteTestRecords()`.
4. Move the cleanup into `CiviTestListener`. Automatically [call setup/teardown functions based on prefixes](https://gist.github.com/totten/d75a721c754acac97347b24f51ed7124). (Any test-trait can define a tear down - eg `Api4TestTrait::tearDown_api4()`).

`#1` is simplest but doesn't go very far to encourage trait decomposition. `#2-4` would go further to encouraging `trait`s, but they're more complicated. `#2` puts the complexity on the test-authors. `#3-#4` put the complexity on the trait-authors.

eileenmcnaughton commented:
I took a look at another place the `require_once` is used and now I think we should rename the `apiv4TestTrait` - it really is just an `EntityProviderTrait` or something like that. There is nothing especially `apiv4` about it. Given that it makes me think a bit differently - ie it's not an equivalent to the `v3Trait` 

I also hit the fact that the test was using https://github.com/civicrm/civicrm-core/pull/25432 yet another class for managing custom fields - which feels like it should be combined with the more heavily used `CRMTraits_Custom_CustomDataTrait` and made generally available to extensions.

We have 10 traits in various stages of maturity in the `tests` directory - so I feel like we should be a bit clearer on what we are working towards than my initial 'this is like apiv3Trait' - although actually implementing feels out of scope of this round
